resharper/resharper-unity/src/CSharp/Daemon/Stages/BurstCodeAnalysis/CallGraph/CallGraphBurstMarksProvider.cs

```diff
@@ -0,0 +1,125 @@
+using System.Collections.Generic;
+using System.Linq;
+using JetBrains.Metadata.Reader.API;
+using JetBrains.Metadata.Reader.Impl;
+using JetBrains.ProjectModel;
+using JetBrains.ReSharper.Daemon.CSharp.CallGraph;
+using JetBrains.ReSharper.Psi;
+using JetBrains.ReSharper.Psi.CSharp.Tree;
+using JetBrains.ReSharper.Psi.Tree;
+using JetBrains.Util;
+
+namespace JetBrains.ReSharper.Plugins.Unity.CSharp.Daemon.Stages.BurstCodeAnalysis.CallGraph
+{
+    [SolutionComponent]
+    public class CallGraphBurstMarksProvider : CallGraphRootMarksProviderBase
+    {
+        private static readonly IClrTypeName TestAttribute = new ClrTypeName("TestAttribute");
+        public CallGraphBurstMarksProvider(ISolution solution)
+            : base(nameof(CallGraphBurstMarksProvider),
+                new CallGraphOutcomingPropagator(solution, nameof(CallGraphBurstMarksProvider)))
+        {
+        }
+
+        public override LocalList<IDeclaredElement> GetRootMarksFromNode(ITreeNode currentNode,
+            IDeclaredElement containingFunction)
+        {
+            var result = new HashSet<IDeclaredElement>();
+            switch (currentNode)
+            {
+                #if JET_MODE_ASSERT
+                case IMethodDeclaration methodDeclaration when methodDeclaration.DeclaredElement is IMethod method && 
+                    method.HasAttributeInstance(TestAttribute, AttributesSource.Self):
+                {
+                    result.Add(method);
+                    break;
+                }
+                #endif
+                case IStructDeclaration structDeclaration
+                    when structDeclaration.DeclaredElement is IStruct @struct &&
+                         @struct.HasAttributeInstance(KnownTypes.BurstCompileAttribute, AttributesSource.Self):
+                {
+                    var visited = new HashSet<ITypeElement>();
+                    var interfaces = new LocalList<ITypeElement>();
+                    var todo = new Stack<ITypeElement>();
+                    visited.Add(@struct);
+                    todo.Push(@struct);
+
+                    while (!todo.IsEmpty())
+                    {
+                        var current = todo.Pop();
+                        foreach (var typeElement in current.GetSuperTypeElements())
```

Comment:
What's about using GetAllSuperTypes and remove while cyle?

Reply:
switched, fixed